state/txindex/kv/kv.go

```diff
@@ -46,13 +59,322 @@ func (txi *TxIndex) Get(hash []byte) (*types.TxResult, error) {
+		it := txi.store.Iterator()
+		defer it.Release()
+		for it.Next() {
+			if !isTagKey(it.Key()) {
+				continue
+			}
+			if strings.Contains(extractValueFromKey(it.Key()), c.Operand.(string)) {
+				hashes = append(hashes, it.Value())
+			}
+		}
+	} else {
+		panic("other operators should be handled already")
+	}
+	return
+}
+
+func (txi *TxIndex) matchRange(r queryRange, startKey []byte) (hashes [][]byte) {
+	it := txi.store.IteratorPrefix(startKey)
+	defer it.Release()
+LOOP:
+	for it.Next() {
+		if !isTagKey(it.Key()) {
+			continue
+		}
+		// no other way to stop iterator other than checking for upperBound
+		switch (r.upperBound).(type) {
+		case int64:
+			v, err := strconv.ParseInt(extractValueFromKey(it.Key()), 10, 64)
+			if err == nil && v == r.upperBound {
+				if r.includeUpperBound {
+					hashes = append(hashes, it.Value())
+				}
+				break LOOP
+			}
+			// XXX: passing time in a ABCI Tags is not yet implemented
+			// case time.Time:
+			// 	v := strconv.ParseInt(extractValueFromKey(it.Key()), 10, 64)
+			// 	if v == r.upperBound {
+			// 		break
+			// 	}
+		}
+		hashes = append(hashes, it.Value())
+	}
+	return
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// Keys
+
+func startKey(c query.Condition, height uint64) []byte {
+	var key string
+	if height > 0 {
+		key = fmt.Sprintf("%s/%v/%d", c.Tag, c.Operand, height)
+	} else {
+		key = fmt.Sprintf("%s/%v", c.Tag, c.Operand)
+	}
+	return []byte(key)
+}
+
+func startKeyForRange(r queryRange, height uint64) []byte {
+	var lowerBound interface{}
+	if r.includeLowerBound {
+		lowerBound = r.lowerBound
+	} else {
+		switch t := r.lowerBound.(type) {
+		case int64:
+			lowerBound = t + 1
+		case time.Time:
+			lowerBound = t.Unix() + 1
+		default:
+			panic("not implemented")
+		}
+	}
+	var key string
+	if height > 0 {
+		key = fmt.Sprintf("%s/%v/%d", r.key, lowerBound, height)
+	} else {
+		key = fmt.Sprintf("%s/%v", r.key, lowerBound)
+	}
+	return []byte(key)
+}
+
+func isTagKey(key []byte) bool {
+	return strings.Count(string(key), tagKeySeparator) == 3
+}
+
+func extractValueFromKey(key []byte) string {
+	parts := strings.SplitN(string(key), tagKeySeparator, 3)
+	return parts[1]
+}
+
+func keyForTag(tag *abci.KVPair, result *types.TxResult) []byte {
+	switch tag.ValueType {
+	case abci.KVPair_STRING:
+		return []byte(fmt.Sprintf("%s/%v/%d/%d", tag.Key, tag.ValueString, result.Height, result.Index))
+	case abci.KVPair_INT:
+		return []byte(fmt.Sprintf("%s/%v/%d/%d", tag.Key, tag.ValueInt, result.Height, result.Index))
+	// case abci.KVPair_TIME:
+	// 	return []byte(fmt.Sprintf("%s/%d/%d/%d", tag.Key, tag.ValueTime.Unix(), result.Height, result.Index))
+	default:
+		panic(fmt.Sprintf("Undefined value type: %v", tag.ValueType))
+	}
+}
+
+///////////////////////////////////////////////////////////////////////////////
+// Utils
+
+func intersect(as, bs [][]byte) [][]byte {
```

Comment:
are these slices guaranteed to be sets (ie. no duplicated elements) ? else we could get duplicates in result. Maybe we can just break from the inner loop once we find a match.

Reply:
see 66ad366a4fd19c13ce6cfb28d6f45e64810f2271. IteratorPrefix prevents from returning the same hash. We may have troubles with the impl if we decide to introduce `OR`

Reply:
How big can these slices get?

I ask because the complexity of this code is currently quadratic `O(n^2)` but could easily be made linear `O(n)` by using a map `map[string]bool` (of which the paradigm `map[string(byte)]` won't incur an allocation) so
```go
aElements := make(map[string]bool)
for i := range as {
   aElements[string(as[i])] = true
}
for j := range bs {
   if _, inA := aElements[string(bs[j])]; inA {
      i = append(i, bs[j])
}
```

Reply:
And btw if they get really big, this is a potential  DOS attack vector as the time is quadratic in the input size.

Reply:
Made a note of this here https://github.com/tendermint/tendermint/issues/907